apps/data-pipeline/sample-program-scraper/src/index.ts

```diff
@@ -0,0 +1,715 @@
+        fall.push(row[0].trim());
+      }
+      if (row.length >= 2 && row[1].trim()) {
+        winter.push(row[1].trim());
+      }
+      if (row.length >= 3 && row[2].trim()) {
+        spring.push(row[2].trim());
+      }
+    }
+
+    transformedProgram.push({
+      year: yearData.year,
+      fall,
+      winter,
+      spring,
+    });
+  }
+
+  return transformedProgram;
+}
+
+// Parse course code into department and course number for functionality lookup
+function parseCourseCode(courseCode: string): { dept: string; num: string } | null {
+  // Match pattern: letters/spaces/& followed by numbers and optional letters
+  const match = courseCode.match(/^([A-Z0-9&\s]+?)(\d+[A-Z]*)$/);
+
+  if (!match) return null;
+
+  return {
+    dept: match[1].trim(),
+    num: match[2],
+  };
+}
+
+async function lookupCourseId(
+  db: ReturnType<typeof database>,
+  courseCode: string,
+): Promise<string | null> {
+  try {
+    const parsed = parseCourseCode(courseCode);
+
+    if (!parsed) {
+      return null;
+    }
+
+    const [foundCourse] = await db
+      .select({ id: course.id })
+      .from(course)
+      .where(and(eq(course.department, parsed.dept), eq(course.courseNumber, parsed.num)))
+      .limit(1);
+
+    return foundCourse?.id ?? null;
+  } catch (error) {
+    logger.warn(`Failed to lookup course: ${courseCode}`, error);
+    return null;
+  }
+}
+
+/*
+  Transforms course code strings to course IDs by looking them up in the database.
+  Falls back to original string if course not found.
+ */
+async function transformCourseCodesToIds(
+  db: ReturnType<typeof database>,
+  scrapedPrograms: {
+    majorId: string;
+    programName: string;
+    variations: {
+      label?: string;
+      sampleProgram: UnvalidatedSampleProgramEntry[];
+      variationNotes: string[];
+    }[];
+  }[],
+): Promise<
+  {
+    majorId: string;
+    programName: string;
+    variations: { label?: string; sampleProgram: SampleProgramEntry[]; variationNotes: string[] }[];
+  }[]
+> {
+  logger.info("Transforming course codes to IDs...");
+
+  const transformed: {
+    majorId: string;
+    programName: string;
+    variations: { label?: string; sampleProgram: SampleProgramEntry[]; variationNotes: string[] }[];
+  }[] = [];
+
+  for (const program of scrapedPrograms) {
+    const transformedVariations: {
+      label?: string;
+      sampleProgram: SampleProgramEntry[];
+      variationNotes: string[];
+    }[] = [];
+
+    for (const variation of program.variations) {
+      const transformedSampleProgram: SampleProgramEntry[] = [];
+
+      for (const year of variation.sampleProgram) {
+        // Transform each term's courses in parallel
+        const [fall, winter, spring] = await Promise.all([
+          Promise.all(
+            year.fall.map(async (code) => {
+              const id = await lookupCourseId(db, code);
+              return id
+                ? { type: "courseId" as const, value: id }
```

Comment:
This code being repeated three times is cumbersome and error-prone. I think a separate helper is warranted which operates on a single code or array of codes, which can be called three times.